plugins/inputs/nsdp/nsdp.go

```diff
@@ -0,0 +1,119 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package nsdp
+
+import (
+	_ "embed"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/tdrn-org/go-nsdp"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type NSDP struct {
+	Address     string          `toml:"address"`
+	DeviceLimit uint            `toml:"device_limit"`
+	Timeout     config.Duration `toml:"timeout"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	conn *nsdp.Conn
+}
+
+func (*NSDP) SampleConfig() string {
+	return sampleConfig
+}
+
+func (n *NSDP) Init() error {
+	if n.Address == "" {
+		n.Address = nsdp.IPv4BroadcastTarget
+	}
+	if n.Timeout == 0 {
+		n.Timeout = config.Duration(2 * time.Second)
+	}
+	return nil
+}
+
+func (n *NSDP) Gather(acc telegraf.Accumulator) error {
+	if n.conn == nil {
+		conn, err := nsdp.NewConn(n.Address, n.Log.Level().Includes(telegraf.Trace))
+		if err != nil {
+			return fmt.Errorf("failed to create connection to address %s: %s", n.Address, err)
+		}
+		conn.ReceiveDeviceLimit = n.DeviceLimit
+		conn.ReceiveTimeout = time.Duration(n.Timeout)
+		n.conn = conn
+	}
+	responses, err := n.conn.SendReceiveMessage(n.newGatherRequest())
+	if err != nil {
+		// Close malfunctioning connection and re-connect on next Gather call
+		n.conn.Close()
+		n.conn = nil
+		return fmt.Errorf("failed to query address %s: %w", n.Address, err)
+	}
```

Comment:
I suggest to do

```suggestion
func (n *NSDP) Start(telegraf.Accumulator) error {
	conn, err := nsdp.NewConn(n.Address, n.Log.Level().Includes(telegraf.Trace))
	if err != nil {
		return fmt.Errorf("failed to create connection to address %s: %s", n.Address, err)
	}
	conn.ReceiveDeviceLimit = n.DeviceLimit
	conn.ReceiveTimeout = time.Duration(n.Timeout)
	n.conn = conn
}

func (n *NSDP) Stop() {
	if n.conn == nil {
		return
	}
	n.conn.Close()
	n.conn = nil
}

func (n *NSDP) Gather(acc telegraf.Accumulator) error {
	if n.conn == nil {
		n.Start(nil)
	}
	responses, err := n.conn.SendReceiveMessage(n.newGatherRequest())
	if err != nil {
		// Close malfunctioning connection and re-connect on next Gather call
		n.Stop()
		return fmt.Errorf("failed to query address %s: %w", n.Address, err)
	}
```

as this will allow Telegraf to automatically close the connection on shutdown.

Reply:
Adapted via 25777c4e4a7e281be320c65c5dd787a4a4a529bb